From 0f96e4b78756deaba77b2d8d486ea5585bd8f652 Mon Sep 17 00:00:00 2001
From: Boudewijn Rempt <boud@valdyas.org>
Date: Thu, 29 Mar 2012 15:32:18 +0200
Subject: [PATCH] Fix inserting table shape in non-english locale

Comparing a translated name with a fixed text isn't going to work...

BUG:297029
BUG:293704

Okay to backport to 2.4?
---
 libs/flake/KoDeferredShapeFactoryBase.h        |    2 ++
 libs/flake/KoShapeFactoryBase.cpp              |    9 ++-------
 sheets/shape/TableShapeDeferredFactory.h       |    2 ++
 sheets/shape/TableShapeFactory.cpp             |    2 +-
 sheets/shape/spreadsheetshape-deferred.desktop |    1 +
 5 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/libs/flake/KoDeferredShapeFactoryBase.h b/libs/flake/KoDeferredShapeFactoryBase.h
index 68bc5e3..7887484 100644
--- a/libs/flake/KoDeferredShapeFactoryBase.h
+++ b/libs/flake/KoDeferredShapeFactoryBase.h
@@ -47,6 +47,8 @@ public:
 
     virtual ~KoDeferredShapeFactoryBase();
 
+    virtual QString deferredPluginName() = 0;
+
     /**
      * This method should be implemented by factories to create a shape that the user
      * gets when doing a base insert. For example from a script.  The created shape
diff --git a/libs/flake/KoShapeFactoryBase.cpp b/libs/flake/KoShapeFactoryBase.cpp
index 116cacc..5c5ec48 100644
--- a/libs/flake/KoShapeFactoryBase.cpp
+++ b/libs/flake/KoShapeFactoryBase.cpp
@@ -242,19 +242,14 @@ void KoShapeFactoryBase::getDeferredPlugin()
     if (d->deferredFactory) return;
 
     const QString serviceType = "Calligra/Deferred";
-    QString query = QString::fromLatin1("(Type == 'Service') and (Name == '%1')").arg(d->deferredPluginName);
-    const KService::List offers = KServiceTypeTrader::self()->query(serviceType, query);
+    const KService::List offers = KServiceTypeTrader::self()->query(serviceType, QString());
     Q_ASSERT(offers.size() > 0);
 
     foreach(KSharedPtr<KService> service, offers) {
-        QString error = 0;  // FIXME: From where does error get a value?
         KoDeferredShapeFactoryBase *plugin = service->createInstance<KoDeferredShapeFactoryBase>(this);
-        if (plugin) {
+        if (plugin && plugin->deferredPluginName() == d->deferredPluginName) {
             d->deferredFactory = plugin;
         }
-        else {
-            kWarning(30003) << "loading plugin" << service->name() << "failed, " << error;
-        }
     }
 
 }
diff --git a/sheets/shape/TableShapeDeferredFactory.h b/sheets/shape/TableShapeDeferredFactory.h
index 2668c2b..981b4d8 100644
--- a/sheets/shape/TableShapeDeferredFactory.h
+++ b/sheets/shape/TableShapeDeferredFactory.h
@@ -30,6 +30,8 @@ public:
     TableDeferredShapeFactory(QObject *parent, const QVariantList&);
     ~TableDeferredShapeFactory();
 
+    virtual QString deferredPluginName() { return "spreadsheetshape-deferred"; }
+
     KoShape *createDefaultShape(KoDocumentResourceManager *documentResources = 0) const;
 private:
     KoShapeFactoryBase *m_stubFactory;
diff --git a/sheets/shape/TableShapeFactory.cpp b/sheets/shape/TableShapeFactory.cpp
index 0cba2a9..98102a6 100644
--- a/sheets/shape/TableShapeFactory.cpp
+++ b/sheets/shape/TableShapeFactory.cpp
@@ -47,7 +47,7 @@ TableShapePlugin::TableShapePlugin(QObject * parent, const QVariantList&)
 
 
 TableShapeFactory::TableShapeFactory()
-    : KoShapeFactoryBase(TableShapeId, i18n("Spreadsheet"), "Spreadsheet Shape")
+    : KoShapeFactoryBase(TableShapeId, i18n("Spreadsheet"), "spreadsheetshape-deferred")
 {
     setToolTip(i18n("Spreadsheet Shape"));
     setIcon("spreadsheetshape");
diff --git a/sheets/shape/spreadsheetshape-deferred.desktop b/sheets/shape/spreadsheetshape-deferred.desktop
index cb47900..da3b679 100644
--- a/sheets/shape/spreadsheetshape-deferred.desktop
+++ b/sheets/shape/spreadsheetshape-deferred.desktop
@@ -39,3 +39,4 @@ Type=Service
 X-KDE-Library=spreadsheetshape-deferred
 X-Flake-MinVersion=4
 X-Flake-PluginVersion=4
+X-DeferredID=spreadsheetshape-deferred
-- 
1.7.9.2

